### ISSUE TITLE
Improve "ignore-order" swiper minibuffer highlighting

### ISSUE THREAD

**Issue by muirdm** [abo-abo/swiper#2167]
I opened this issue to recap the problem and get feedback on the best approach to fix. There are two related causes of bad minibuffer highlighting:
 
1. `swiper--re-builder` in some cases related to "^" munges the regex parts into a single regex string. This breaks "!" parts for all regexes and messes up highlighting for "ignore-order". I think we should get away from ever turning regex parts into a single regex since that always breaks "!" groups, and often breaks "ignore-order" matching and/or highlighting.

2. The fact that `re-builder` is set to something else (`swiper--re-builder`), the highlight function gets set to "default" instead of "ignore-order".

I think the underlying issue is that swiper uses its own re-builder. It is hard to maintain proper behavior for the various builders with the `swiper--re-builder` black box in between swiper and ivy.

Some ideas to fix:

1. Change `ivy-read` so you can separate "label" (what user sees) from "value" (what gets returned and what the filtering applies to). In the swiper case, the label would be `123 foo.bar()` and the value would just be `foo.bar()`. That way the normal regex builder could be used and we could get rid of `swiper--re-builder`.

2. Fix `swiper--re-builder` maintain the separate regex parts and then tweak `ivy--reset-state` to set `ivy--highlight-function` to
```
(or
  (assq ivy--regex-function ivy-highlight-functions-alist)
  (assq (ivy-alist-setting ivy-re-builders-alist) ivy-highlight-functions-alist)
  #'ivy--highlight-default))
```
which would more or less fix things, but maybe not as comprehensively.

I also realized regarding `swiper-isearch`: does it need to use `swiper--re-builder` at all since it doesn't include line numbers? Maybe I should give up on `swiper` and focus on `swiper-isearch`?

**Comment by abo-abo** [comment]
Can you add a scenario when the minibuffer highlighting is wrong?

**Comment by muirdm** [comment]
It looks like issue 1) was resolved in 33fa992aebe9605a3130d221543a232dafce5ab9 since `swiper--re-builder` no longer forces builder output into a single string.

Issue 2) can be reproduced by setting builder to `ivy--regex-ignore-order` and swipering for `foo bar` in buffer with contents:
```
foo bar
```

The problem is the minibuffer uses the first match face for both "foo" and "bar". However, now that 1) is fixed, it is easy enough to make ivy-read choose the right highlighter when called by swiper.
